src/pool/auction.ts

```diff
@@ -0,0 +1,254 @@
+    const scaledAuction: AuctionData = {
+      block: this.auctionData.block,
+      bid: new Map(),
+      lot: new Map(),
+    };
+    let lotModifier: bigint;
+    let bidModifier: bigint;
+    const fillBlockDelta = BigInt(scaleBlock - this.auctionData.block);
+    const perBlockScaler = BigInt(50000);
+    const percentFillScaled = BigInt(scalePercent * 100000);
+    if (fillBlockDelta <= 200n) {
+      lotModifier = fillBlockDelta * perBlockScaler;
+      bidModifier = FixedMath.SCALAR_7;
+    } else {
+      lotModifier = FixedMath.SCALAR_7;
+      if (fillBlockDelta < 400n) {
+        bidModifier = FixedMath.SCALAR_7 - (fillBlockDelta - 200n) * perBlockScaler;
+      } else {
+        bidModifier = 0n;
+      }
+    }
+    for (const [assetId, amount] of Array.from(this.auctionData.lot)) {
+      const scaledLot = FixedMath.mulFloor(
+        FixedMath.mulFloor(amount, percentFillScaled, FixedMath.SCALAR_7),
+        lotModifier,
+        FixedMath.SCALAR_7
+      );
+      if (scaledLot > 0n) {
+        scaledAuction.lot.set(assetId, BigInt(scaledLot));
+      }
+    }
+
+    for (const [assetId, amount] of Array.from(this.auctionData.bid)) {
+      const scaledBid = FixedMath.mulCeil(
+        FixedMath.mulCeil(amount, percentFillScaled, FixedMath.SCALAR_7),
+        bidModifier,
+        FixedMath.SCALAR_7
+      );
+      if (scaledBid > 0) {
+        scaledAuction.bid.set(assetId, BigInt(scaledBid));
+      }
+    }
+    return new ScaledAuction(
+      this.user,
+      this.type,
+      scaledAuction.bid,
+      scaledAuction.lot,
+      this.auctionData.block,
+      scaleBlock,
+      filled
+    );
+  }
+}
+
+export class ScaledAuction {
+  type: AuctionType;
+  user: string;
+  bid: Map<string, i128>;
+  lot: Map<string, i128>;
+  startBlock: number;
+  scaleBlock: number;
+  filled: boolean;
+  // the hash of the transaction that filled the auction
+  hash: string | undefined;
+  // the timestamp of the auction creation
+  timestamp: string | undefined;
+
+  constructor(
+    user: string,
+    type: AuctionType,
+    bid: Map<string, i128>,
+    lot: Map<string, i128>,
+    startBlock: number,
+    scaleBlock: number,
+    filled: boolean,
+    hash?: string,
+    timestamp?: string
+  ) {
+    this.user = user;
+    this.type = type;
+    this.bid = bid;
+    this.lot = lot;
+    this.startBlock = startBlock;
+    this.scaleBlock = scaleBlock;
+    this.filled = filled;
+    this.hash = hash;
+    this.timestamp = timestamp;
+  }
+
+  /**
+   * Creates an array of scaled auctions from events representing the state of the auctions at the current ledger
+   * @param events - The events to create the scaled auctions from
+   * @param backstopId - The address of the backstop
+   * @param currLedger - The current ledger
+   * @returns An array of scaled auctions
+   */
+  static fromEvents(events: PoolEvent[], backstopId: string, currLedger: number): ScaledAuction[] {
```

Comment:
Should this be on `Auction` rather than `ScaledAuction`?

This needs to be rebuilt every block now, instead of having a list of auctions, and allowing them to be scaled individually.